dynamiq/memory/backend/in_memory.py

```diff
@@ -0,0 +1,56 @@
+from dynamiq.memory.backend.base import MemoryBackend
+from dynamiq.prompts import Message
+
+
+class InMemoryError(Exception):
+    """Base exception class for InMemory backend errors."""
+
+    pass
+
+
+class InMemory(MemoryBackend):
+    """In-memory implementation of the memory storage backend."""
+
+    name = "InMemory"
+
+    def __init__(self):
+        """Initializes the in-memory storage."""
+        self.messages: list[Message] = []
+
+    def add(self, message: Message):
+        """Adds a message to the in-memory list."""
+        try:
+            self.messages.append(message)
+        except Exception as e:
+            raise InMemoryError(f"Error adding message to InMemory backend: {e}") from e
+
+    def get_all(self) -> list[Message]:
+        """Retrieves all messages from the in-memory list."""
+        return sorted(self.messages, key=lambda msg: msg.metadata.get("timestamp", 0))  # Sort by timestamp
+
+    def search(self, query: str = None, limit: int = None, filters: dict = None) -> list[Message]:
+        """Searches for messages, applying optional query and/or filters."""
+        limit = limit or self.config.search_limit
+        matching_messages = self.messages
+
+        if query:
+            matching_messages = [msg for msg in matching_messages if query.lower() in msg.content.lower()]
```

Comment:
Could you please explain this use case a little bit? We are searching for the exact inclusion of a query in the previous messages.

My question is: Do we want to extract the exact duplicate of the previous user query input? If yes, how will this information help us?

Or do we want to extract answers where this query was mentioned? Are we 100% sure that the answers include the needed input query in the exact match formulation?

I think it would be beneficial to have the ability for semantic retrieval with high similarity confidence, and later with the option to assess with an LLM or cross-encoder if the memory contains the same information or similar but different information that we want to reuse.

Reply:
@MaksymAI there is inmemory implementation, without embeddings and not vector store, so just simple search

Reply:
@MaksymAI we have semantic search only upon vector stores backend